### ISSUE TITLE
Addition of "none" type reduction in loss functions.

### ISSUE THREAD

**Issue by kartikdutt18** [mlpack/mlpack#2211]
Hi everyone, I wanted to know if it would be okay to add support for "none" type in loss function. Currently we return mean or sum in loss functions (Generally, not true everywhere). Using none type users could have the following use cases:

1. Masking: None type reduction is useful for masking (each point will have a loss associated with it).

2. Allow user flexibility: Software should be more user friendly i.e. user should be able to see how he got the given loss. Allowing support for none type will give that opportunity.

3. Make mlpack similar to Pytorch, TensorFlow or Keras. They all support none type reduction. 
    To quote from summer of code ideas,

> mlpack has a great ANN module but when compared to Tensorflow or Pytorch it still lags behind in terms of features and flexibility. Hence it would be great if the missing features can be hunted down and implemented/fixed. 

If this is okay, I would love to work on the same. Thanks.

**Comment by kartikdutt18** [comment]
Hi @zoq, what do you think?

**Comment by kartikdutt18** [comment]
Lets try to keep this one open.

**Comment by kartikdutt18** [comment]
Thanks @favre49, Do you mind sharing your opinion on this?

**Comment by favre49** [comment]
Sorry for not replying sooner, didn't notice this issue.

I agree that this would be a good addition to mlpack, I rthink it was discussed on the IRC last year (though I don't remember the specifics).

> 1.Overload Forward Function.
 >  2. In place Changes (To input).
    Ideally: change function return type based on reduction parameter (Looking for ways to implement the same).

I'll need to go through the code again to give more advice on this. I'll get back to you on this in a day or so. From memory, option 2 doesn't seem possible, but I think we could do option 1? Since you posted this issue, have you gotten any insight on a way forward?

**Comment by kartikdutt18** [comment]
> I'll need to go through the code again to give more advice on this.
Great.
I think the best option would be overloaded something like this`forward(input, target, output)` but then it might not mesh well with the API. Maybe there is another way?
